crates/amaru-ledger/src/governance/ratification.rs

```diff
@@ -407,22 +366,40 @@ impl<'distr> RatificationContext<'distr> {
         let span = tracing::Span::current();
 
         let empty = BTreeMap::new();
-
         let (dreps_votes, cc_votes, pool_votes) =
             partition_votes(self.votes.get(id).unwrap_or(&empty));
 
+        // NOTE: because ratification is an expensive operation, and something that we *may* have
+        // to replay due to rollbacks, it's important to do the least amount of work possible.
+        //
+        // If the CC doesn't approve, then the majority of actions can be considered invalid and
+        // there's no need to check for DReps nor SPO votes. The following code is thus written in
+        // a way that the next governance body is only consulted should the previous one be "no".
+
         let cc_approved = self.is_accepted_by_constitutional_committee(proposal, cc_votes);
+
         span.record("approved.committee", cc_approved);
 
-        let spos_approved =
-            self.is_accepted_by_stake_pool_operators(proposal, pool_votes, stake_distribution);
-        span.record("approved.pools", spos_approved);
+        if cc_approved {
+            let spos_approved =
+                self.is_accepted_by_stake_pool_operators(proposal, pool_votes, stake_distribution);
+
+            span.record("approved.pools", spos_approved);
 
-        let dreps_approved =
-            self.is_accepted_by_delegate_representatives(proposal, dreps_votes, stake_distribution);
-        span.record("approved.dreps", dreps_approved);
+            if spos_approved {
```

Comment:
Would it not be useful to have some unit tests for this logic? Not sure how complicated this would be to set up, but having a manifestation of this "shortcut" logic would help our future selves.

Reply:
I've been trying to see how to best test the "plumbing" at the top, but I've been struggling so far. So my strategy has been to make it *as thin as possible* and have the lower parts well tested.